Source/driver/Castro.cpp

```diff
@@ -954,6 +954,56 @@ Castro::initData ()
 
        }
 
+       // it is not a requirement that the problem setup defines the
+       // temperature, so we do that here _and_ ensure that we are
+       // within any small limits
+       computeTemp(S_new, cur_time, S_new.nGrow());
+
+       int init_failed_rho = 0;
+       int init_failed_T = 0;
+
+#ifdef _OPENMP
+#pragma omp parallel reduction(+:init_failed_rho,init_failed_T)
+#endif
+       for (MFIter mfi(S_new, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+         {
+           const Box& bx = mfi.tilebox();
+
+           auto S_arr = S_new.array(mfi);
+
+           Real lsmall_temp = small_temp;
+           Real lsmall_dens = small_dens;
+
+           int* init_failed_rho_d = AMREX_MFITER_REDUCE_SUM(&init_failed_rho);
+           int* init_failed_T_d = AMREX_MFITER_REDUCE_SUM(&init_failed_T);
+
+           AMREX_PARALLEL_FOR_3D(bx, i, j, k,
+           {
+
+             // if the problem tried to initialize a thermodynamic
+             // state that is at or below small_temp, then we abort.
+             // This is dangerous and we should recommend a smaller
+             // small_temp
+             if (S_arr(i,j,k,UTEMP) < lsmall_temp * 1.001) {
+               *init_failed_T_d += 1;
```

Comment:
In order to be thread-safe on the device, these += statements need to be done in a thread-safe way for the CUDA build (this is fine for OpenMP). See how we use `deviceReduceSum` in `sum_utils.cpp`.